### PR TITLE
Distributed tracing and async segments

### PR DESCRIPTION
**Pull request by bobbyrward**
This should take care of most of the missing features from the c-sdk as well as add in some rudimentary support for running futures in segments.

I think this is all I'll be adding for now and I'll be running this in our test environment for the time being to make sure I didn't make any dumb unsafe mistakes.

Thanks again for getting all this together in the first place.  That's made getting our rust code into newrelic so much easier.

### CODE REVIEW

File: src/app.rs
Comment (bobbyrward) [inline src/app.rs]: I knew I was forgetting something...

**Comment by sd2k** [comment]
This looks great, thanks for the PR!

As I mentioned I can't really test this any more since I don't have a New Relic subscription, but I'll add a few comments where I can.

File: src/app.rs
Comment (sd2k) [inline src/app.rs:71]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: This could use a bit more detail, even if it's just a link to the C SDK docs.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs:124]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: This could use a bit more detail, even if it's just a link to the C SDK docs.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: This could use a bit more detail, even if it's just a link to the C SDK docs.

File: src/app.rs
Comment (sd2k) [inline src/app.rs]: Needs a proper doc comment.

File: src/futures.rs
Comment (sd2k) [inline src/futures.rs:229]: This is really elegant, I like it! Could you document the fields on the struct so they appear in the API docs?

File: src/lib.rs
Comment (sd2k) [inline src/lib.rs:75]: Could you make these additions in the README as well?

**Comment by bobbyrward** [comment]
That's understandable.  We're pretty tied to newrelic at this point so if you want help maintaining this or to pass ownership, let me know.

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs]: I think the description for this is copy/pasted from above!

It also looks like we can use the following attribute so that documentation marks feature-gated structs/methods:

```
#[cfg(feature = "distributed_tracing")]
#[cfg_attr(docsrs, doc(cfg(feature = "distributed_tracing")))]
pub fn distributed_trace(&self) -> Result<String> {
...
```

This can be done for the `async` structs/traits in futures.rs too.

File: src/app.rs
Comment (sd2k) [reply src/app.rs]: This should be gated behind the `distributed_tracing` feature too, so users don't think they're getting tracing without the feature being enabled.

File: src/futures.rs
Comment (sd2k) [inline src/futures.rs:106]: Completely down to preference but this could be shortened to...

```suggestion
        SegmentedFuture {
            inner: self,
            segment: to_trans
                .get_transaction()
                .map(|t| segment::Segment::custom(t, name, category)),
        }
```

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs:301]:
```suggestion
    #[cfg(feature = "distributed_tracing")]
    #[cfg_attr(docsrs, doc(cfg(feature = "distributed_tracing")))]
```

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs:335]:
```suggestion
#[cfg(feature = "distributed_tracing")]
#[cfg_attr(docsrs, doc(cfg(feature = "distributed_tracing")))]
```

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs:339]:
```suggestion
#[cfg(feature = "distributed_tracing")]
#[cfg_attr(docsrs, doc(cfg(feature = "distributed_tracing")))]
```

**Review by sd2k — CHANGES_REQUESTED** [review]
I've added some minor comments, mostly related to doc comments.

The only other major thing is an example of using the new `async` and `distributed_tracing` features. In particular I don't really know how distributed tracing works - I guess you send the payload generated by `Segment.distributed_trace()` in any HTTP requests to other services?

File: src/futures.rs
Comment (sd2k) [reply src/futures.rs:106]: Hmm, this suggestion is only replacing the last line even though I highlighted all of the above block! Never mind...

File: src/app.rs
Comment (sd2k) [inline src/app.rs]:
```suggestion
///     )
///     .unwrap()
```

File: src/app.rs
Comment (sd2k) [inline src/app.rs]:
```suggestion
/// let license_key = std::env::var("NEW_RELIC_LICENSE_KEY")
```

**Comment by sd2k** [comment]
> That's understandable. We're pretty tied to newrelic at this point so if you want help maintaining this or to pass ownership, let me know.

Yep, I think it makes a lot of sense to add you as a maintainer at least if you're happy with that, and potentially pass ownership later if I still can't really use it :+1:

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs]: This should either utilize a new `Utf8Error` variant of the `newrelic::Error` type and return something like

```
        c_str
            .to_str()
            .map(ToString::to_string)
            .map_err(Error::Utf8Error)
```

...or should just continue using `.expect()` and return a `String` (instead of `Result<String>`), since we're getting a HTTP safe payload anyway.

File: src/segment.rs
Comment (sd2k) [inline src/segment.rs:297]:
```suggestion
    /// # }
    /// # Ok::<(), newrelic::Error>(())
```

This doctest fails without this due to the `?`.

**Comment by bobbyrward** [comment]
I'll work on improving the docs.  I'm not really familiar with rustdoc so this gives me a chance to get familiar with how all of this works.

File: src/futures.rs
Comment (bobbyrward) [reply src/futures.rs:229]: The struct is left intentionally opaque here.  I think some doc tests would make it easier to understand.  I'll need to pull tokio into the dev dependencies but that should be fine.

**Comment by bobbyrward** [comment]
I added but don't really understand the line: `#[cfg_attr(docsrs, doc(cfg(feature = "distributed_tracing")))]`

**Comment by sd2k** [comment]
To be honest I just took it straight from tokio - I think it's used by docs.rs to indicate that certain methods are only available with a feature enabled: https://docs.rs/tokio/0.2.11/tokio/runtime/struct.Builder.html#method.threaded_scheduler

**Comment by bobbyrward** [comment]
Ohhh, ok.  That makes sense.  I didn't see any difference in local docs and couldn't tell what it was doing.

File: src/futures.rs
Comment (sd2k) [reply src/futures.rs:229]: Yep fair enough. Doc tests look good now!

File: src/segment.rs
Comment (sd2k) [reply src/segment.rs:297]: Ah I can see you changed this to remove the `?`. Since users will need to use the headers perhaps it'd be better to re-add the `?`, and add an `    /// # Ok::<(), newrelic::Error>(())` below the existing doctest lines, which will be a bit more representative of real usage. (There's an example of this in the [documentation tests](https://doc.rust-lang.org/rustdoc/documentation-tests.html#using--in-doc-tests) Rust docs.

Comment (bobbyrward) [reply src/segment.rs:297]: Huh.  I thought it had changed this.  Any objection to taking out the `if false` on this so that doesn't happen again?

Comment (sd2k) [reply src/segment.rs:297]: No objection - the tests don't all currently pass without the New Relic key anyway. Hopefully I'll get CI working soon...

Comment (bobbyrward) [reply src/segment.rs:297]: I verified it's working now.

I had also previously changed the distributed_trace call not to return a result since, it shouldn't ever fail to decode a base64 string as utf8.

**Review by sd2k — APPROVED** [review]

**Comment by sd2k** [comment]
LGTM! Happy to merge if this is working for you with New Relic.

**Comment by bobbyrward** [comment]
This has been running for a little over 2 weeks now without any restarts or loss of monitoring.   I think it's good.

**Comment by sd2k** [comment]
Great news. I've merged, updated the changelog, and released version 0.2.2 to crates.io :tada: 

Thanks again for the contributions!

**Pull request merged**
